core/src/main/java/org/apache/iceberg/DeleteFilesTable.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+
+/**
+ * A {@link Table} implementation that exposes a table's delete files as rows.
+ */
+public class DeleteFilesTable extends BaseFilesTable {
+
+  DeleteFilesTable(TableOperations ops, Table table) {
+    this(ops, table, table.name() + ".delete_files");
+  }
+
+  DeleteFilesTable(TableOperations ops, Table table, String name) {
+    super(ops, table, name);
+  }
+
+  @Override
+  public TableScan newScan() {
+    return new DeleteFilesTableScan(operations(), table(), schema());
+  }
+
+  @Override
+  MetadataTableType metadataTableType() {
+    return MetadataTableType.DELETE_FILES;
+  }
+
+  public static class DeleteFilesTableScan extends BaseFilesTableScan {
+    private final Schema fileSchema;
+
+    DeleteFilesTableScan(TableOperations ops, Table table, Schema fileSchema) {
+      super(ops, table, fileSchema, MetadataTableType.DELETE_FILES);
+      this.fileSchema = fileSchema;
+    }
+
+    private DeleteFilesTableScan(TableOperations ops, Table table, Schema schema, Schema fileSchema,
+                           TableScanContext context) {
+      super(ops, table, schema, fileSchema, context, MetadataTableType.DELETE_FILES);
+      this.fileSchema = fileSchema;
+    }
+
+    @Override
+    protected TableScan newRefinedScan(TableOperations ops, Table table, Schema schema, TableScanContext context) {
+      return new DeleteFilesTableScan(ops, table, schema, fileSchema, context);
+    }
+
+    @Override
+    protected CloseableIterable<FileScanTask> planFiles(
+        TableOperations ops, Snapshot snapshot, Expression rowFilter,
+        boolean ignoreResiduals, boolean caseSensitive, boolean colStats) {
+      CloseableIterable<ManifestFile> filtered = filterManifests(snapshot.deleteManifests(), rowFilter, caseSensitive);
+
+      String schemaString = SchemaParser.toJson(schema());
+      String specString = PartitionSpecParser.toJson(PartitionSpec.unpartitioned());
+      Expression filter = ignoreResiduals ? Expressions.alwaysTrue() : rowFilter;
+      ResidualEvaluator residuals = ResidualEvaluator.unpartitioned(filter);
+
+      // Data tasks produce the table schema, not the projection schema and projection is done by processing engines.
+      // This data task needs to use the table schema, which may not include a partition schema to avoid having an
+      // empty struct in the schema for unpartitioned tables. Some engines, like Spark, can't handle empty structs in
+      // all cases.
+      return CloseableIterable.transform(filtered, manifest ->
+          new DeleteManifestReadTask(ops.io(), ops.current().specsById(),
+              manifest, schema(), schemaString, specString, residuals));
+    }
+  }
+
+  static class DeleteManifestReadTask extends BaseFileScanTask implements DataTask {
```

Comment:
What about making the existing task for reading manifests a little bit more generic rather than introducing a similar class? We will reuse it also in ALL_DELETE_FILES.

Technically, you could move `ManifestReadTask` to `BaseFilesTable`.

```
@Override
public CloseableIterable<StructLike> rows() {
  return CloseableIterable.transform(manifestEntries(), file -> (StructLike) file);
}

private CloseableIterable<? extends ContentFile<?>> manifestEntries() {
  switch (manifest.content()) {
    case DATA:
      return ManifestFiles.read(manifest, io, specsById).project(schema);
    case DELETES:
      return ManifestFiles.readDeleteManifest(manifest, io, specsById).project(schema);
    default:
      throw new IllegalArgumentException("Unsupported manifest content type:" + manifest.content());
  }
}
```

Reply:
Yea this is a good idea and makes the final metadata table scan class very minimal, done